### PR TITLE
Introduce guard controller for static pods

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Have independent deployment and PDB controllers stitched together to form guard controller

### CODE REVIEW

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+func New(
+	name string,
+	manifest []byte,
+	recorder events.Recorder,
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	deployInformer appsinformersv1.DeploymentInformer,
+	optionalInformers []factory.Informer,
+	optionalManifestHooks []ManifestHookFunc,
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:65]: doc the difference between this and the deploymenthooks.  The construction here is good and I understand why, but be sure to explain it godoc for others.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	// TODO: Collapse this into optional deployment hook.
+	optionalManifestHooks []ManifestHookFunc
+	// Optional hook functions to modify the Deployment.
+	// If one of these functions returns an error, the sync
+	// fails indicating the ordinal position of the failed function.
+	// Also, in that scenario the Degraded status is set to True.
+	optionalDeploymentHooks []DeploymentHookFunc
+}
+
+func New(
```
Comment (deads2k) [inline pkg/operator/deploymentcontroller/deployment_controller.go:57]: `NewDeploymentController` : the golang recommendations are wrong.

File: pkg/operator/staticpod/controllers.go [hunk 3]
```diff
@@ -138,6 +150,17 @@ func (b *staticPodOperatorControllerBuilder) WithPruning(command []string, stati
 	return b
 }
 
+func (b *staticPodOperatorControllerBuilder) WithGuard(deploymentAssetFunc func(string) []byte, deploymentFile string, pdbAssetFunc func(string) []byte, pdbFile string) Builder {
```
Comment (deads2k) [inline pkg/operator/staticpod/controllers.go:153]: The guard logically requires the `infraLister configv1listers.InfrastructureLister`, so add `infraLister configv1listers.InfrastructureLister` as an arg to this method please.

File: pkg/operator/staticpod/controllers.go [hunk 4]
```diff
@@ -262,6 +285,27 @@ func (b *staticPodOperatorControllerBuilder) ToControllers() (manager.Controller
...
+	if len(b.deploymentManifest) > 0 && len(b.pdbManifest) > 0 {
+		manager.WithController(pdbcontroller.New(b.staticPodOperatorClient,
+			b.kubeClient,
+			b.kubeInformers,
+			b.eventRecorder,
+			b.infraLister,
+			b.pdbManifest,
+			b.operandNamespace), 1)
+		nodeInformer := b.kubeInformers.InformersFor("").Core().V1().Nodes()
+		manager.WithController(deploymentcontroller.New("static-pod",
```
Comment (deads2k) [inline pkg/operator/staticpod/controllers.go:299]: name as something with guard.  `StaticPodGuardDeployment`?

File: pkg/operator/pdbcontroller/pdb_controller.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	clusterTopology      configv1.TopologyMode
+	targetNamespace      string
+	pdbManifest          []byte
+	factory              *factory.Factory
+	eventRecorder        events.Recorder
+}
+
+var _ factory.Controller = &Controller{}
+
+func New(operatorClient v1helpers.OperatorClient,
```
Comment (deads2k) [inline pkg/operator/pdbcontroller/pdb_controller.go:39]: this should accept a name too.

Comment (deads2k) [inline pkg/operator/pdbcontroller/pdb_controller.go:39]: Given this `New` method a decent name, the golang recommendation of `New` is nearly always wrong.

Comment (deads2k) [inline pkg/operator/pdbcontroller/pdb_controller.go:39]: I'm ok with this controller as a short term solution if @sttts and  @soltysh want to land in 4.8.  If they prefer waiting until 4.9, please develop the idea of a precondition which can decorate a static resource controller.

`NewStaticResourceController().WithConditionalResource(conditionFunc, manifest, deleteConditionFunc)`.  If the deleteConditionFunc returns true, delete the resource, if the conditionFunc returns true create the resource. Deletion trumps creation.  By default, deletionCondtionFunc is !conditionFunc.

**Comment by deads2k** [comment]
separate out the deploymentcontroller refactor from the rest and get storage to review.  This is the direction I'm thinking of.

File: pkg/operator/deploymentcontroller/deployment_controller.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/operator/deploymentcontroller/deployment_controller.go:67]: Sure, I explained it [here](https://github.com/openshift/library-go/pull/1095/files#diff-abb679a52a6ea86c466b84d87793ee37682c0e4b6a111234e69cb684d7ec3157R44)

**Review by ravisantoshgudimetla — COMMENTED** [review]
Split the commits and addressed your comments. 

cc @bertinatto @jsafrane for reviewing f606e021e58fd2f11d250b405c98532e207519a8 and c68f4286f1fac3ae786980d6533a19e185af6931

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go [hunk 6]
```diff
@@ -101,187 +61,11 @@ func NewCSIDriverControllerServiceController(
...
-		pairs = append(pairs, []string{"${KUBE_RBAC_PROXY_IMAGE}", kubeRBACProxy}...)
-	}
-
-	// Cluster ID
-	pairs = append(pairs, []string{"${CLUSTER_ID}", clusterID}...)
-
-	// Log level
-	logLevel := loglevel.LogLevelToVerbosity(spec.LogLevel)
-	pairs = append(pairs, []string{"${LOG_LEVEL}", strconv.Itoa(logLevel)}...)
+	optionalInformers = append(optionalInformers, configInformer.Config().V1().Infrastructures().Informer())
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:66]: I was initially thinking of moving tests around but I realized that the hooks(both replicas and placeholder hooks are needed by you). The replicas function you have different from the replicas function we want in common controller, so I did not end up moving the tests but I added tests for the hooks I added in deployment.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	progressingCondition := opv1.OperatorCondition{
+		Type:   c.name + opv1.OperatorStatusTypeProgressing,
+		Status: opv1.ConditionFalse,
+	}
+
+	if ok, msg := isProgressing(deployment); ok {
+		progressingCondition.Status = opv1.ConditionTrue
+		progressingCondition.Message = msg
+		progressingCondition.Reason = "Deploying"
+	}
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply pkg/operator/deploymentcontroller/deployment_controller.go:166]: We're planning to land the current PR in 4.9 timeframe. I hope you're fine with it. If not, I'll split this PR into 2
- For the refactor of deployment controller
- One for guard controller
as I want to address David's [comment ](https://github.com/openshift/library-go/pull/1095#discussion_r646860219)

[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/deploymentcontroller/deployment_controller.go:166]: no, leave it as it is, it goes the right direction. We can update it when we know what we want (the linked PR has some issues).

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:66]: I mean now there is lot of code in `deploymentcontroller` (like the conditions setting) that is currently tested only in `csidrivercontrollerservicecontroller`. This is IMO bad, the code should be tested in package where it's written. Can you please at least *copy* the tests to `deploymentcontroller`, without any hooks, and we can then prune our CSI tests in a subsequent PR?

**Comment by jsafrane** [comment]
/assign @bertinatto 
for CSI details. LGTM from my side.

File: pkg/operator/deploymentcontroller/deployment_controller.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+// DeploymentHookFunc is a hook function to modify the Deployment.
+type DeploymentHookFunc func(*opv1.OperatorSpec, *appsv1.Deployment) error
+
+// ManifestHookFunc is a hook function to modify the manifest in raw format.
+type ManifestHookFunc func(*opv1.OperatorSpec, []byte) ([]byte, error)
+
+// DeploymentController is a generic controller that manages a deployment
+
+//
+// <name>Available: indicates that the deployment controller  was successfully deployed and at least one Deployment replica is available.
```
Comment (bertinatto) [inline pkg/operator/deploymentcontroller/deployment_controller.go:34]: nit: empty comment, double spaces

File: pkg/operator/pdbcontroller/pdb_controller.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+	if haTopologyMode != configv1.HighlyAvailableTopologyMode {
+		return nil
+	}
+
+	if err := pdbc.ensureGuardingPDB(ctx, syncCtx.Recorder()); err != nil {
+		return err
+	}
+
+	if err != nil {
```
Comment (bertinatto) [inline pkg/operator/pdbcontroller/pdb_controller.go:100]: This won't be reached because you're returning `err` above.

Also, I don't think you need to set the Degraded condition yourself, as it's set when you return an error.

File: pkg/operator/pdbcontroller/pdb_controller.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("PDBControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(pdbc.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "PDBControllerDegraded",
```
Comment (bertinatto) [inline pkg/operator/pdbcontroller/pdb_controller.go:115]: Is there any reason you're setting Degraded here? Typically you would only need to set Available.

File: pkg/operator/pdbcontroller/pdb_controller.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	return updateErr
+}
+
+func (pdbc *PDBController) ensureGuardingPDB(ctx context.Context, recorder events.Recorder) error {
+	guardPDB := resourceread.ReadPodDisruptionBudgetV1OrDie(pdbc.pdbManifest)
+
+	_, _, err := resourceapply.ApplyPodDisruptionBudget(pdbc.kubeClient.PolicyV1(), recorder, guardPDB)
+	if err != nil {
+		klog.Errorf("Failed to verify/apply %s pdb, error %w", guardPDB.Name, err)
+		return err
```
Comment (bertinatto) [inline pkg/operator/pdbcontroller/pdb_controller.go:128]: nit: no need to log the error if we're returning it

**Review by bertinatto — COMMENTED** [review]
Added a few comments.

I'm not opposed to merging 8ec123cbd3988ccb08cdcc6080163ed4972ef981, but is it possible to use the generic Deployment controller with a brand new controller first?

Also, besides the `CSIDriverControllerServiceController` we also have a very similar controller that deploys a DaemonSet. Will there be a `DaemonSetController`?

**Review by ravisantoshgudimetla — COMMENTED** [review]
> Added a few comments.
> 
> I'm not opposed to merging [8ec123c](https://github.com/openshift/library-go/commit/8ec123cbd3988ccb08cdcc6080163ed4972ef981), but is it possible to use the generic Deployment controller with a brand new controller first?
> 
> Also, besides the `CSIDriverControllerServiceController` we also have a very similar controller that deploys a DaemonSet. Will there be a `DaemonSetController`?

@jsafrane and @bertinatto - thank you both for the reviews. I have addressed your comments. PTAL.

As discussed over slack, we're in the process of using generic deployment in all control-plane components. But we would like to get this PR merged before we go using it. Yes, we also have plans of having a DSController and Stscontroller in future. So, let us know if you already have a DSController, you can contribute it to library-go directly.

File: pkg/operator/resource/resourceapply/generic.go [hunk 12]
```diff
@@ -52,6 +52,18 @@ type ApplyResult struct {
...
+// ConditionalFunction provides needed dependency for a resource on another condition instead of blindly creating
+// a resource. This conditional function can also be used to delete the resource when not needed
+type ConditionalFunction func() bool
+
+// ResourceConditionalMap maps a file to conditional functions that can be used to create or delete a resource in the
+// given file.
+type ResourceConditionalMap struct {
+	File                  string
+	DeleteConditionalFunc ConditionalFunction
+	CreateConditionalFunc ConditionalFunction
```
Comment (jsafrane) [inline pkg/operator/resource/resourceapply/generic.go:64]: What if both functions return `true`? IMO it would be better to have a single function that returns eg. `ResourceAction`, which can be either `"Create"` or `"Delete"`

Comment (ravisantoshgudimetla) [reply pkg/operator/resource/resourceapply/generic.go:64]: https://github.com/openshift/library-go/pull/1095/commits/59ec4d268990a31f4c0604ce2dc2c509ab5cac2c#diff-b91f253f2414046d3d35cc689c04410dfc00c7b8130a953ac59926118f9b8423R119 - Deletion trumps creation.

Comment (jsafrane) [reply pkg/operator/resource/resourceapply/generic.go:64]: Yes, I saw it. But what benefit do the two function have over a single function? IMO it makes the code more error prone.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 13]
```diff
@@ -102,25 +114,40 @@ func (c *StaticResourceController) WithIgnoreNotFoundOnCreate() *StaticResourceC
 	return c
 }
 
+// WithConditionalResource makes the controller react to changes in dependent object
+// If the deleteConditionFunc returns true, delete the resource, if the conditionFunc returns true create the resource.
+// Deletion trumps creation. By default, deletionConditionFunc is !conditionFunc.
+// TODO: This can be made multi-resource aware as well.
+func (c *StaticResourceController) WithConditionalResource(createConditionFunc resourceapply.ConditionalFunction,
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:121]: `createConditionFn, deleteConditionFn, files ...string`

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 14]
```diff
@@ -102,25 +114,40 @@ func (c *StaticResourceController) WithIgnoreNotFoundOnCreate() *StaticResourceC
...
+	file string, deleteConditionFunc resourceapply.ConditionalFunction) *StaticResourceController {
+	rcm := resourceapply.ResourceConditionalMap{
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:123]:
```go
conditionalManifests := (read them)
conditionalApplyFn := func() (directResourceResults, error){
  if deletionCondtionFn(){
    return resourceapply.DeleteAll(ctx, c.clients, syncContext.Recorder(), conditionalManifests), nil
  }
  if createConditionFn(){
    return resourceapply.ApplyDirectly(ctx, c.clients, syncContext.Recorder(), conditionalManifests), nil
  }
  return nil, nil
}
c.ConditionalApplyFns = append(conditionalApplyFn)

```

Then you add these below the existing ApplyDirectly in Sync.

Comment (ravisantoshgudimetla) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go]: @deads2k - I slightly tweaked your suggestion and applied the changes. PTAL when you find time.

File: pkg/operator/resource/resourceapply/apiextensions.go [hunk 15]
```diff
@@ -40,3 +40,20 @@ func ApplyCustomResourceDefinitionV1(ctx context.Context, client apiextclientv1.
...
+func DeleteCustomResourceDefinitionV1(ctx context.Context, client apiextclientv1.CustomResourceDefinitionsGetter, recorder events.Recorder, required *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, bool, error) {
+	existing, err := client.CustomResourceDefinitions().Get(ctx, required.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, false, nil
+		} else {
+			return nil, false, err
+		}
+	}
+	err = client.CustomResourceDefinitions().Delete(ctx, existing.Name, metav1.DeleteOptions{})
```
Comment (deads2k) [inline pkg/operator/resource/resourceapply/apiextensions.go:53]: you can save a call by simply calling delete and ignoring NotFound.

Comment (ravisantoshgudimetla) [reply pkg/operator/resource/resourceapply/apiextensions.go:53]: I believe you don't want me to do a get? I am fine with it.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 16]
```diff
@@ -219,6 +265,24 @@ func (c StaticResourceController) Sync(ctx context.Context, syncContext factory.
 			continue
 		}
 	}
+	// Handle conditional resources that need to be created or deleted.
+	if c.conditionalResource != nil && len(c.conditionalResource.conditionalFiles) > 0 {
+		var conditionalResourceResults []resourceapply.ApplyResult
+		if c.conditionalResource.shouldDelete {
+			conditionalResourceResults = resourceapply.DeleteAll(ctx, c.clients, syncContext.Recorder(), c.manifests, c.conditionalResource.conditionalFiles...)
+		} else {
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:273]: you shouldn't create unless the create function returns true.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 17]
```diff
@@ -102,6 +111,43 @@ func (c *StaticResourceController) WithIgnoreNotFoundOnCreate() *StaticResourceC
 	return c
 }
 
+func shouldDelete(createConditionalFunc, deleteConditionalFunc resourceapply.ConditionalFunction) bool {
+	shouldDelete := false
+	if createConditionalFunc == nil && deleteConditionalFunc == nil {
+		shouldDelete = false
+	} else {
+		if createConditionalFunc != nil && createConditionalFunc() {
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:119]: delete should only be controlled by the deletionfunc.  Create should only be controlled by the creation func.

if the create and delete func both report true, an error should be reported (do this in sync)

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 18]
```diff
@@ -219,6 +265,24 @@ func (c StaticResourceController) Sync(ctx context.Context, syncContext factory.
 			continue
 		}
 	}
+	// Handle conditional resources that need to be created or deleted.
+	if c.conditionalResource != nil && len(c.conditionalResource.conditionalFiles) > 0 {
+		var conditionalResourceResults []resourceapply.ApplyResult
+		if c.conditionalResource.shouldDelete {
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:271]:
```go
shouldDelete := false
if shouldDeleteFn != nil && deleteFn(){
  shouldDelete = true
}

shouldCreate != false
if shouldCreateFn != nil && shouldCreateFn(){
  shouldCreate = true
}

if shouldCreate && shouldDelete{
  errors = append(errors, wtf)
}

if shouldDelete{
  deleteAll
} else if shouldCreate{
  createAll
}

accumulateresults.
```

**Comment by jsafrane** [comment]
I have another use case for StaticResourceController removing objects it manages: OLM / optional operators. Here I need StaticResourceController to remove a finalizer on a CR when everything has been removed. I think it is not possible with current callbacks - there is no way how to tell if StaticResourceController has finished removing everything it wanted or not.

Perhaps I miss some context, does static pod controller need a similar signal to tell that PDB has been removed?

**Comment by ravisantoshgudimetla** [comment]
> I have another use case for StaticResourceController removing objects it manages: OLM / optional operators. Here I need StaticResourceController to remove a finalizer on a CR when everything has been removed. I think it is not possible with current callbacks - there is no way how to tell if StaticResourceController has finished removing everything it wanted or not.


The goal in this case is to create, delete static resource based on conditions not the other way around but I am wondering if your use-case can be solved by simple chaining. How do you create CR? Using static resource controller?

**Comment by jsafrane** [comment]
My CR will be created / deleted by user. What do you mean by the "simple chaining"?

File: pkg/operator/v1helpers/interfaces.go [hunk 19]
```diff
@@ -19,7 +19,7 @@ type OperatorClient interface {
 }
 
 type StaticPodOperatorClient interface {
-	OperatorClient
+	OperatorClientWithFinalizers
```
Comment (soltysh) [inline pkg/operator/v1helpers/interfaces.go:22]: Why this change, I'm not seeing it being used here nor in https://github.com/openshift/cluster-kube-controller-manager-operator/pull/543

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 20]
```diff
@@ -102,6 +112,46 @@ func (c *StaticResourceController) WithIgnoreNotFoundOnCreate() *StaticResourceC
...
+//}
+
+func (c *conditionalResource) shouldDelete() {
+
+}
+
+// WithConditionalResource makes the controller react to changes in dependent object
+// If both createConditional and deleteConditional are true, it will be considered an
+// error during sync.
+func (c *StaticResourceController) WithConditionalResource(createConditionalFunc resourceapply.ConditionalFunction,
```
Comment (soltysh) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:145]: Maybe `WithConditionalResources`?

File: pkg/operator/resource/resourceapply/generic.go [hunk 21]
```diff
@@ -223,6 +227,135 @@ func ApplyDirectly(ctx context.Context, clients *ClientHolder, recorder events.R
...
+			} else {
+				_, result.Changed, result.Error = DeleteStorageVersionMigration(ctx, clients.migrationClient, recorder, t)
+			}
+		case *unstructured.Unstructured:
+			if clients.dynamicClient == nil {
+				result.Error = fmt.Errorf("missing dynamicClient")
+			} else {
+				_, result.Changed, result.Error = DeleteKnownUnstructured(ctx, clients.dynamicClient, recorder, t)
+			}
+		default:
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply pkg/operator/resource/resourceapply/generic.go:349]: To be in sync with `ApplyDirectly` method.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go (hunk 20 above)
Comment (ravisantoshgudimetla) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go]: Sure. I changed it.

File: pkg/operator/v1helpers/interfaces.go (hunk 19 above)
Comment (ravisantoshgudimetla) [reply pkg/operator/v1helpers/interfaces.go:22]: It is being used in https://github.com/openshift/library-go/pull/1095/commits/1df083fadac8328066bb845a4d6049a2870742c5#diff-d0ce926df996bfa572967a0d2a1d011bbfc84cd41e065b957328285647b17cecR293, technically deploymentController started using it and I had to modify the methods to be implemented in staticPodOperatorClient :( but I think in the long run this good as we can provide OperatorClientWithFinalizers for all the staticpodClients.

**Review by ravisantoshgudimetla — COMMENTED** [review]
@soltysh - I answered your questions. PTAL

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 22]
```diff
@@ -58,9 +58,19 @@ type StaticResourceController struct {
 
 	eventRecorder events.Recorder
 
-	factory          *factory.Factory
-	restMapper       meta.RESTMapper
-	categoryExpander restmapper.CategoryExpander
+	factory *factory.Factory
+
+	conditionalResource *conditionalResource
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:63]: this should be a slice

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 23]
```diff
@@ -219,6 +242,36 @@ func (c StaticResourceController) Sync(ctx context.Context, syncContext factory.
...
+		if c.conditionalResource.deleteConditional != nil && c.conditionalResource.deleteConditional() {
+			shouldDelete = true
+		}
+
+		shouldCreate := false
+		if c.conditionalResource.createConditional != nil && c.conditionalResource.createConditional() {
+			shouldCreate = true
+		}
+
+		if shouldCreate && shouldDelete {
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:258]: switch/case please

File: pkg/operator/staticresourcecontroller/static_resource_controller.go (hunk 22 above)
Comment (ravisantoshgudimetla) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go:63]: The [fields](https://github.com/openshift/library-go/pull/1095/files#diff-b91f253f2414046d3d35cc689c04410dfc00c7b8130a953ac59926118f9b8423R71) that this structure is holding is a slice. Do we still need make this slice?

Comment (deads2k) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go:63]: based on your current types, `conditionalResources []conditionalResource`.  I should be able to call `WithConditionalResources` multiple times.

Comment (ravisantoshgudimetla) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go:63]: Ok. I can make that change.

**Review by ravisantoshgudimetla — COMMENTED** [review]
@soltysh - Updated and modified the units. PTAL

**Review by soltysh — APPROVED** [review]
The changes here looks reasonable, but before merging I'd like to see how this is being used (updated version) in https://github.com/openshift/cluster-kube-controller-manager-operator/pull/543 Especially such that we don't have to introduce yet another pod but rather re-use one of the pre-existing.
